flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/GSFileSystem.java

```diff
@@ -39,17 +45,67 @@
+
+        // follow the same rules as for the Hadoop connector, i.e.
+        // 1) only use service credentials at all if Hadoop
+        // "google.cloud.auth.service.account.enable" is true (default: true)
+        // 2) use GOOGLE_APPLICATION_CREDENTIALS as location of credentials, if supplied
+        // 3) use Hadoop "google.cloud.auth.service.account.json.keyfile" as location of
+        // credentials, if supplied
+        // 4) use no credentials
+
+        // store any credentials we are to use, here
+        Optional<String> credentialsPath = Optional.empty();
+
+        // only look for credentials if service account support is enabled
+        Configuration hadoopConfig = getHadoopFileSystem().getConf();
+        boolean enableServiceAccount =
+                hadoopConfig.getBoolean("google.cloud.auth.service.account.enable", true);
+        if (enableServiceAccount) {
+
+            // load google application credentials, and then fall back to
+            // "google.cloud.auth.service.account.json.keyfile" from Hadoop
+            credentialsPath = Optional.ofNullable(System.getenv("GOOGLE_APPLICATION_CREDENTIALS"));
+            if (credentialsPath.isPresent()) {
+                LOGGER.info(
+                        "Recoverable writer is using GOOGLE_APPLICATION_CREDENTIALS at {}",
+                        credentialsPath.get());
+            } else {
+                credentialsPath =
+                        Optional.ofNullable(
+                                hadoopConfig.get("google.cloud.auth.service.account.json.keyfile"));
+                credentialsPath.ifPresent(
+                        path ->
+                                LOGGER.info(
+                                        "Recoverable writer is using credentials from Hadoop at {}",
+                                        path));
+            }
+        }
 
-        // create the Google storage service instance
-        Storage storage = StorageOptions.getDefaultInstance().getService();
+        // construct the storage instance, using credentials if provided
+        Storage storage;
+        if (credentialsPath.isPresent()) {
+            LOGGER.info(
+                    "Creating GSRecoverableWriter using credentials from {}",
+                    credentialsPath.get());
+            try (FileInputStream credentialsStream = new FileInputStream(credentialsPath.get())) {
+                GoogleCredentials credentials = GoogleCredentials.fromStream(credentialsStream);
+                storage =
+                        StorageOptions.newBuilder()
+                                .setCredentials(credentials)
+                                .build()
+                                .getService();
+            }
+        } else {
+            LOGGER.info("Creating GSRecoverableWriter using no credentials");
+            storage = StorageOptions.newBuilder().build().getService();
+        }
```

Comment:
I'd suggest to minimize things we do in the `if-else` branches as follow:
```
        // construct the storage instance, using credentials if provided
        StorageOptions.Builder storageOptionBuilder = StorageOptions.newBuilder();
        if (credentialsPath.isPresent()) {
            LOGGER.info(
                    "Creating GSRecoverableWriter using credentials from {}",
                    credentialsPath.get());
            try (FileInputStream credentialsStream = new FileInputStream(credentialsPath.get())) {
                GoogleCredentials credentials = GoogleCredentials.fromStream(credentialsStream);
                storageOptionBuilder.setCredentials(credentials);
            }
        } else {
            LOGGER.info("Creating GSRecoverableWriter using no credentials");
        }

        // create the GS blob storage wrapper
        GSBlobStorageImpl blobStorage =
                new GSBlobStorageImpl(storageOptionBuilder.build().getService());
```

Reply:
Good suggestion, I'll make that change.

Reply:
Done in https://github.com/apache/flink/pull/18489/commits/6d137f72504470a6be9d971b7e1bba1d8614cffa (and moved to `ConfigUtils`).